Feature task history

irinaudovyk opened the pull request.

irinaudovyk committed:
- added new index.js for comment`s endpoints;
- added new 'commentsCounter' field in task model and TaskForm;
- added TaskHistory and TaskComment models and registered in container.js;
- added post remove function for comments in Task scema;
- added a new icon of comments to taskMetrics.html;
- added form for creating comments in task.html;
- added comments to task.html;

irinaudovyk committed:
-added a new directive taskHistory;
-added task counter icon to taskMetrics.html;
- added Detail link to comments ;
- added comment saving on enter+shift;
-added recognize links in comment text;

irinaudovyk committed:
-deleted index.js for comments;
-created new index.js for history (history writers and comments);
- added TaskStatus,TaskSpenttime,TaskMetrics, TaskDescription, TaskDeveloper,TaskComplexity models and StatusWriter, MetricsWriter,SpenttimeWriter,DescriptionWriter,DeveloperWriter,ComplexityWriter clases, included to container;
-deleted field createdAt in TaskHistorySchema;
- added HistoryService, included to container;
-added commentsCounter directive;
- created templates for different type of history message;
- added filter buttons to taskHistory.html and taskHistory.js ;
-created taskHistoryLabel directive;
-created historyInfoService;
created TaskHistory service;

eugenetsukanov commented on app/api/history/index.js:
```diff
@@ -0,0 +1,50 @@
+                if (err) {
+                    return next(err);
+                }
+                if (!messages) {
```
looks like .find returns array if no errors

eugenetsukanov commented on app/api/history/index.js:
```diff
@@ -0,0 +1,50 @@
+                if (!messages) {
+                    return res.sendStatus(404);
+                }
+                res.status(200).send(messages);
```
not good idea, cause you try to send text, we use json instead

eugenetsukanov commented on app/api/task/index.js:
```diff
@@ -23,7 +23,8 @@ module.exports = function (app) {
         field("tagsList").array(),
-        field("archived").trim()
+        field("archived").trim(),
+        field("commentsCounter").isNumeric()
```
do we really need it here ? we shouldn't allow to set counters

eugenetsukanov commented on app/models/task.js:
```diff
@@ -47,6 +49,16 @@ module.exports = function (mongoose) {
     });
 
+    TaskSchema.post('remove', function (doc) {
+        var TaskComment = mongoose.model('TaskComment');
```
could you remove all history here ?

eugenetsukanov commented on public/tracker/modules/task/directives/commentsCounter/commentsCounter.js:
```diff
@@ -0,0 +1,20 @@
+            templateUrl: 'tracker/modules/task/directives/commentsCounter/commentsCounter.html',
+            controller: function ($scope) {
+
+                $scope.toggleTaskHistory = function ($event) {
```
looks like you need clean up this code, not used in directive template

eugenetsukanov commented on public/tracker/modules/task/directives/taskHistory/taskHistory.html:
```diff
@@ -0,0 +1,39 @@
+            <span ng-if="message.user"><strong>{{message.user.first}}{{message.user.last}}</strong></span>
+            <span class="text-muted"
+                  style="font-size: 10px;"> at {{message.updatedAt | date : 'HH:mm:ss yyyy/MM/dd'}}</span>
+            <div ng-include="getInclude(message._type)" class="history-message"></div>
```
looks like would be better to move 'task-history-label', span with user name and date to include templates

eugenetsukanov commented on public/tracker/modules/task/directives/taskHistory/taskHistory.js:
```diff
@@ -0,0 +1,84 @@
+                    var allMessages;
+
+                    $scope.getInclude = function (messageType) {
+                        var historyType = _.find(historyInfoService, function (historyType) {
```
looks like if we move all rendering to templates we no need historyInfoService, and this service can be remove

eugenetsukanov commented on public/tracker/modules/task/directives/taskHistory/taskHistory.js:
```diff
@@ -0,0 +1,84 @@
+
+                    function updateCommentCounter() {
+                        $scope.task.commentsCounter++;
+                        $scope.task.$update({taskId: $scope.task._id});
```
this is responsibility of server side, it ($scope.task.commentsCounter++;) can be easy hacked

eugenetsukanov commented on public/tracker/modules/task/task.services.js:
```diff
@@ -297,5 +297,54 @@ angular
             }
         ]
     })
+    .factory('historyInfoService', function () {
```
about template urls looks like can be move to history directive

eugenetsukanov reviewed (changes requested):
please could you check comments

irinaudovyk committed:
- added updateCommentsCounter and notifyUsers to index.js of task history;
-cleaned up;
-deleted 'commentsCounter' field from TaskForm;
- added task post 'remove' for TaskHistory;
- added updateCommentsCounter method to TaskService;
- deleted commentsCounter directive;
- created taskComment directive;
- creted templites for diffrent types of task history;
- deleted historyInfoService in task.services.js;
- added sockets on event 'comment.save';
-deleted taskHistoryLabel directive;

irinaudovyk committed:
Merge branch 'dev' into feature-task-history-merge2

Conflicts:
	app/services/TaskService.js

irinaudovyk committed:
- deleted spaces in task.html;
